### PR TITLE
Speed up updating user parameter value

### PR DESCRIPTION
**Pull request by qnkhuat**
3.a of https://github.com/metabase/metabase/issues/45780

- Add an index on user_parameter_value(parameter_id, user_id) -- with this alone I think we can reduce a lot of contention and avoid deadlock
- Reduce # of db calls to update user_param_value. We could get it down to 2 db calls if we want to, see this [comment](https://github.com/metabase/metabase/pull/46615#issuecomment-2274988705)
- Fix N+1 in :last_used_paramater_values hydration where N is # of dashboard (though I don't think we batched hydrate this anywhere)

Result of viewing the ecormece insights dashboard with 4 filters
Screenshot 2024-08-14 at 11 09 46

master: 333 DB calls
this branch: 316 db calls

### CODE REVIEW

**Comment by qnkhuat** [comment]
There is another way to reduce the # db calls further, which is by using insert on conflict.
But this requires adding a unique constraint on (user_id, dashboard_id, parameter_id) and I'm not sure we whether we should do that.

I'd let the @metabase/admin-webapp and @dpsutton make the call here as their team developed this feature.

File: src/metabase/models/user_parameter_value.clj [hunk 1]
```diff
@@ -51,25 +51,46 @@
...
+(mu/defn batched-upsert!
+  "Delete param with nil value and upsert the rest."
+  [user-id         :- ms/PositiveInt
+   dashboard-id    :- ms/PositiveInt
+   param-id->value :- [:map-of :string :any]]
+  (let [to-delete (filter (comp nil? second) param-id->value)
+        to-upsert (filter (comp some? second) param-id->value)]
+    (t2/with-transaction [_conn]
+      (doseq [[param-id value] to-upsert]
+        (or (pos? (t2/update! :model/UserParameterValue {:user_id      user-id
```
Comment (qnkhuat) [inline src/metabase/models/user_parameter_value.clj:63]: if we add a unique constraint on `parameter_id, dashboard_id, user_id` we can do this in one statement
```clojure
(t2/query {:insert-into :user_parameter_value
                  :values      (map (fn [[param-id value]]
                                      {:user_id      user-id
                                       :dashboard_id dashboard-id
                                       :parameter_id param-id
                                       :value        value})
                                    to-upsert)
                  :on-conflict  [:user_id :dashboard_id :parameter_id]
                  :do-update-set :value})
```

File: src/metabase/models/user_parameter_value.clj [hunk 2]
```diff
@@ -51,25 +51,46 @@
...
-                                       (fn [{:keys [parameter_id value]}]
-                                         [parameter_id value])
-                                       :model/UserParameterValue
-                                       :user_id user-id
-                                       :dashboard_id dashboard-id
-                                       :parameter_id [:in all-parameter-ids]))))]
-      (map
-       (fn [{dashboard-id :id :as dashboard}]
-         (let [param-ids (mapv :id (:parameters dashboard))]
-           (assoc dashboard :last_used_param_values (select-keys (last-used-values dashboard-id) param-ids)))) dashboards))
```
Comment (qnkhuat) [inline src/metabase/models/user_parameter_value.clj:74]: there was an N+1 here where N is # of the dashboard. Though

File: src/metabase/models/user_parameter_value.clj [hunk 3]
```diff
@@ -30,46 +30,46 @@
...
                                                    :dashboard_id dashboard-id
-                                                   :parameter_id parameter-id
+                                                   :parameter_id param-id
                                                    :value        value})))
-    (t2/delete! :model/UserParameterValue
-                :user_id      user-id
-                :dashboard_id dashboard-id
-                :parameter_id parameter-id)))
+      (when (seq to-delete)
+        (t2/delete! :model/UserParameterValue
```
Comment (qnkhuat) [inline src/metabase/models/user_parameter_value.clj:51]: This group delete will reduce many db calls if you view a dashboard without any param value

File: src/metabase/models/user_parameter_value.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/models/user_parameter_value.clj:42]: Is there any reason not to add it?

File: src/metabase/models/user_parameter_value.clj [hunk 4]
```diff
@@ -30,46 +30,46 @@
...
+     :last_used_param_values
+     (fn [] ;; return a map of {dashboard-id {parameter-id value}}
+       (let [upvs (t2/select :model/UserParameterValue
+                             :dashboard_id [:in (map :id dashboards)]
+                             :user_id user-id)]
+         (as-> upvs result
+           (group-by :dashboard_id result)
+           (update-vals result (fn [upvs]
+                                 (into {}
+                                       (map (juxt :parameter_id :value) upvs)))))))
```
Comment (piranha) [inline src/metabase/models/user_parameter_value.clj:73]:
```suggestion
         (-> (group-by :dashboard_id upvs)
             (update-vals (fn [upvs]
                            (into {}
                                  (map (juxt :parameter_id :value) upvs)))))))
```

File: src/metabase/models/user_parameter_value.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/models/user_parameter_value.clj:42]: note: you'll have to write `merge into` for h2 and `insert ... update` for mysql though :)

**Review by sansolovyov — APPROVED** [review]
Nice work! No comments aside from trivial stuff.

**Comment by qnkhuat** [comment]
@piranha I've updated to implement the suggestion I made [here](https://github.com/metabase/metabase/pull/46615#issuecomment-2274988705)

**Review by sansolovyov — APPROVED** [review]
I think that's cool! 

I wonder if you have any appetite to put a check for a count of queries around the API? :)

**Review by escherize — APPROVED** [review]
Thanks @qnkhuat, this looks great. I was not sure about the mysql `batched-upsert!` implementation, is there a good reason to not use the new constraint there?

**Comment by qnkhuat** [comment]
Unfortunately, adding the unique index will make it easy to hit race conditions if you have one filter connected to multiple parameters.

I can reproduce easily with one param that connects to 3 cards. 
I think to fully resolve this we need to re-think our schema for user_parameter_value. For now I'll reverse the change to add the unique index.

**Pull request merged**